### PR TITLE
[go-metro] mem fixes. Don't let bookkeeping get too big for flows.

### PR DESCRIPTION
**Pull request by truthbk**

### CODE REVIEW

File: ddtypes.go [hunk 1]
```diff
@@ -15,7 +15,9 @@ type TCPKey struct {
 }
 
 const (
-	CHAN_DEPTH = 10
+	CHAN_DEPTH      = 10
+	FLUSH_IVAL      = 600
```
Comment (truthbk) [inline ddtypes.go]: Wondering if this value may be a bit too small? Just wanted to be conservative.

File: reporter.go [hunk 2]
```diff
@@ -74,12 +75,22 @@ func (r *Client) Report() error {
...
 		select {
 		case key := <-r.flows.Expire:
 			r.flows.Delete(key)
 			log.Infof("Flow expired: [%s]", key)
 		case <-ticker.C:
+			flush := false
+			now := time.Now().Unix()
+
+			runtime.ReadMemStats(&memstats)
+			pct := float64(memstats.Alloc) / float64(memstats.Sys)
```
Comment (olivielpeau) [inline reporter.go]: qq:`memstats.Sys` reflects the `bytes obtained from system` by the runtime, is it what you want to know, or would you rather use the total memory of the system?

File: reporter.go [hunk 3]
```diff
@@ -74,12 +75,22 @@ func (r *Client) Report() error {
...
+			if pct >= FORCE_FLUSH_PCT { //memory out of control
+				flush = true
```
Comment (olivielpeau) [inline reporter.go]: We probably want to log something when this happens

File: reporter.go [hunk 4]
```diff
@@ -121,6 +132,9 @@ func (r *Client) Report() error {
 						log.Debugf("Reported successfully on: %v", k)
 					}
 				}
+				if flush || (now-flow.LastFlush) > FLUSH_IVAL {
```
Comment (olivielpeau) [inline reporter.go]: where is `now-flow` defined?

**Comment by olivielpeau** [comment]
Had a couple comments, other than that it looks good to me!

File: ddtypes.go (hunk 1 above)
Comment (truthbk) [inline ddtypes.go]: We'll keep it as is :+1:

File: reporter.go [hunk 5]
```diff
@@ -72,14 +103,36 @@ func (r *Client) Report() error {
 
 	log.Infof("Started reporting.")
 
+	memsize, err := memorySize()
+	if err != nil {
+		log.Warnf("Error getting memory size. Relying on OOM to keep process in check.")
```
Comment (olivielpeau) [inline reporter.go]: fwiw we could log the `err` too

File: reporter.go [hunk 6]
```diff
@@ -72,14 +103,36 @@ func (r *Client) Report() error {
...
+			runtime.ReadMemStats(&memstats)
+			if memsize > 0 {
+				pct = float64(memstats.Alloc) / float64(memsize)
+			} else {
+				pct = 0
+			}
+
+			if pct >= FORCE_FLUSH_PCT { //memory out of control
+				flush = true
+				log.Warnf("Forcing flush - memory consumption above maximum allowed system usage: %v %", pct*100)
```
Comment (olivielpeau) [inline reporter.go]: I _think_ it's safer to escape the percent sign here, i.e. `%%`

**Comment by olivielpeau** [comment]
Added 2 small comments, once they're addressed feel free to squash and merge!

**Pull request merged**